src/decoder/stream.rs

```diff
@@ -131,63 +131,94 @@ pub type DeflateReader<'r, R> = flate2::read::ZlibDecoder<&'r mut SmartReader<R>
 ///
 
 /// Reader that decompresses LZW streams
-pub struct LZWReader {
-    buffer: io::Cursor<Vec<u8>>,
+pub struct LZWReader<R: Read> {
+    compressed_buf: Box<[u8]>,
+    compressed_buf_length: usize,
+    compressed_bytes_read: usize,
+
+    reader: R,
```

Comment:
Isn't that more or less a duplicate of `BufReader<Take<R>>`?

Reply:
The frustrating thing is that `BufReader` has no "grow buffer" functionality. `BufReader::fill_buf` only guarantees that the internal buffer will have at least one byte in the internal buffer (assuming there is more input), but weezl may need multiple bytes of input to successfully advance the internal state.

Reply:
Hm? That doesn't sound right. As long as there's an output buffer and a non-empty code buffer it should advance state _somehow_. It also does [partial refills](https://github.com/image-rs/lzw/blob/master/src/decode.rs#L997-L1002) of the internal codeword bit buffer if the input is too short for a full word.

Reply:
If you have a test to the contrary–that is lack of progress with a non-empty input **and** output buffer– I'd regard that as a bug to `weezl`.

Reply:
The documentation of LzwStatus::NoProgress says "No bytes were read or written and no internal state advanced."

Reply:
But looking in more detail, it seems the tests that were triggering NoProgress were doing so by passing an empty input buffer

Reply:
There is no minimum input length to advance though. As long as you provide bytes and decoding isn't stuck on writing, it will advance input state and consume bytes. There's no reason that requires filling the input buffer to a _any_ minimum size.